### PR TITLE
First batch of fixes for test_other with LLVM backend

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -2556,14 +2556,14 @@ def do_binaryen(target, asm_target, options, memfile, wasm_binary_target,
...
-    if shared.Settings.SIDE_MODULE:
+  if shared.Settings.SIDE_MODULE:
+    if not shared.Settings.WASM_BACKEND:
       wso = shared.WebAssembly.make_shared_library(final, wasm_binary_target)
       # replace the wasm binary output with the dynamic library. TODO: use a specific suffix for such files?
       shutil.move(wso, wasm_binary_target)
       if not DEBUG:
         os.unlink(asm_target) # we don't need the asm.js, it can just confuse
-      sys.exit(0) # and we are done.
+    sys.exit(0) # and we are done.
```
Comment (jgravelle-google) [inline emcc.py]: So we now exit 0 if SIDE_MODULE and WASM_BACKEND, where before we would ignore SIDE_MODULE and do compilation normally?
Neither of those seem great, we're silently not doing what the user asked.
How about change the `if not WASM_BACKEND` to `assert not WASM_BACKEND`? Or is there some other reason we don't do that

File: tests/test_other.py [hunk 2]
```diff
@@ -833,6 +837,8 @@ def test(args, expected, moar_expected=None):
       out, err = Popen([PYTHON, EMCC, 'src.c'] + args, stderr=PIPE).communicate()
       self.assertContained(expected, run_js(self.in_dir('a.out.js'), stderr=PIPE, full_output=True, assert_returncode=None))
       print('in asm.js')
+      if is_wasm_backend():
```
Comment (jgravelle-google) [inline tests/test_other.py]: Should be `self.is_wasm_backend()`?

File: tests/test_other.py [hunk 3]
```diff
@@ -3784,14 +3799,14 @@ def test_bad_function_pointer_cast(self):
...
-                Popen(cmd).communicate()
+                run_process(cmd)
                 output = run_js('a.out.js', stderr=PIPE, full_output=True, assert_returncode=None)
                 if emulate_casts:
                   # success!
                   self.assertContained('Hello, world.', output)
                 else:
                   # otherwise, the error depends on the mode we are in
-                  if wasm and (relocate or emulate_fps):
+                  if self.is_wasm_backend() or (wasm and (relocate or emulate_fps)):
```
Comment (jgravelle-google) [inline tests/test_other.py]: Wasm backend always mismatches function signatures here?

File: tests/test_other.py [hunk 4]
```diff
@@ -7957,25 +7993,29 @@ def test_binaryen_debug(self):
...
-          asm2wasm_line = [x for x in err.split('\n') if 'asm2wasm' in x][0]
-          asm2wasm_line = asm2wasm_line.strip() + ' ' # ensure it ends with a space, for simpler searches below
-          print('|' + asm2wasm_line + '|')
-          assert expect_dash_g == (' -g ' in asm2wasm_line)
-          assert expect_emit_text == (' -S ' in asm2wasm_line)
-          if expect_emit_text:
-            text = open('a.out.wast').read()
-            assert ';;' in text, 'must see debug info comment'
-            assert 'hello_world.cpp:6' in text, 'must be file:line info'
+          if self.is_wasm_backend():
```
Comment (jgravelle-google) [inline tests/test_other.py]: How about `if not self.is_wasm_backend()` and skip the pass/else

File: tests/test_other.py [hunk 5]
```diff
@@ -8192,16 +8238,17 @@ def test_wasm_targets(self):
...
+      self.assertContained('output file "%s" has a wasm suffix, but we cannot emit wasm by itself, except as a dynamic library' % f, process.stderr)
     # side modules do allow a wasm target
     for opts, target in [([], 'a.out.wasm'), (['-o', 'lib.wasm'], 'lib.wasm')]:
       # specified target
-      print(target)
+      print('building: ' + target)
       self.clear()
-      subprocess.check_call([PYTHON, EMCC, path_from_root('tests', 'hello_world.cpp'), '-s', 'WASM=1', '-s', 'SIDE_MODULE=1'] + opts)
-      assert b'dylink' in open(target, 'rb').read()
+      cmd = [PYTHON, EMCC, path_from_root('tests', 'hello_world.cpp'), '-s', 'WASM=1', '-s', 'SIDE_MODULE=1'] + opts
```
Comment (jgravelle-google) [inline tests/test_other.py]: Probably worth just leaving the `cmd` inline here

File: tests/test_other.py [hunk 6]
```diff
@@ -8337,12 +8384,15 @@ def test_single_file(self):
...
+      expect_wasm = wasm_enabled
       expect_emterpretify_file = emterpreter_file_enabled
       expect_meminit = (meminit1_enabled and not wasm_enabled) or (wasm_enabled and asmjs_fallback_enabled)
       expect_success = not (emterpreter_file_enabled and single_file_enabled)
-      expect_wasm = wasm_enabled
-      expect_wast = debug_enabled and wasm_enabled
+      expect_asmjs_code = asmjs_fallback_enabled and wasm_enabled and not self.is_wasm_backend()
+      expect_wast = debug_enabled and wasm_enabled and not self.is_wasm_backend()
+
+      if self.is_wasm_backend() and (asmjs_fallback_enabled or emterpreter_enabled or not wasm_enabled):
```
Comment (jgravelle-google) [inline tests/test_other.py]: `asmjs_fallback_enabled` is redundant here (can only be true if `self.is_wasm_backend()` is false, which would always make this if false)

File: tools/shared.py [hunk 7]
```diff
@@ -38,8 +38,8 @@ class FatalError(Exception):
   pass
 
 
-def exit_with_error(*args):
-  logging.error(*args)
+def exit_with_error(msg, *args):
```
Comment (jgravelle-google) [inline tools/shared.py]: Is the explicit `msg` here to make it clearer that the first argument is special, or to better avoid calling this improperly (with 0 arguments)?
Either way, nice.

File: tools/shared.py [hunk 8]
```diff
@@ -1343,6 +1342,9 @@ def verify_settings():
     if Settings.EMTERPRETIFY:
       exit_with_error('emcc: EMTERPRETIFY is not is not supported by the LLVM wasm backend')
```
Comment (jgravelle-google) [inline tools/shared.py]: Nit not from this change: "EMTERPRETIFY is not is not supported" should just be "is not supported", unless we really want to emphasize that it's not supported and we mean that.

File: emcc.py (hunk 1 above)
Comment (jgravelle-google) [reply emcc.py]: Or if we disallow SIDE_MODULE and WASM_BACKEND from being set together in Settings itself (at the end of this diff, in shared.py), do we need the `if not WASM_BACKEND` here at all?

Comment (sbc100) [reply emcc.py]: This was my first attempt at gettting SIDE_MODULE to work with WASM_BACKEND.  Basically we want to early exit here when SIDE_MODULE is used so we don't go onto great .js stuff.   The make_shared_library() call doesn't currently work with wasm backend, but we exiting is still better than continuing.  I could just revert this.. but actually I think its best to remove the WASM_BACKEND check here instead, and left verify_settings() to the checking..

File: tests/test_other.py (hunk 3 above)
Comment (sbc100) [reply tests/test_other.py]: Yes, all modes have the mismatch but the specific error varies.  It seems that the wasm backend always produces this one.

File: tools/shared.py (hunk 7 above)
Comment (sbc100) [reply tools/shared.py]: This was part of a separate change to actually re-raise in debug mode.. but its seemed like an innocuous thing to squeeze in there.

File: tests/test_other.py [hunk 9]
```diff
@@ -7728,7 +7754,8 @@ def test_binaryen_opts(self):
           print(args, 'js opts:', expect_js_opts, 'only-wasm:', expect_only_wasm, '   ', ' '.join(cmd))
           err = run_process(cmd, stdout=PIPE, stderr=PIPE).stderr
           assert expect_js_opts == ('applying js optimization passes:' in err), err
-          assert expect_only_wasm == ('-emscripten-only-wasm' in err and '--wasm-only' in err), err # check both flag to fastcomp and to asm2wasm
+          if not self.is_wasm_backend():
```
Comment (kripken) [inline tests/test_other.py]: the whole test is no-wasm-backend, though, why is this needed? preparation for eventually enabling parts of it?

Comment (jgravelle-google) [reply tests/test_other.py]: Way I read it, the test is currently relying on asm2wasm-specific behavior. But this part should never happen in wasm backend, so it makes sense to make that expectation explicit.

Comment (kripken) [reply tests/test_other.py]: I see, sounds good.

Comment (sbc100) [reply tests/test_other.py]: Yes, I initially tried to make this test pass with wasm backend, but gave up at some point. But I think this could still be useful going forward. I could revert this part I suppose.

Comment (kripken) [reply tests/test_other.py]: No need to revert, lgtm as is.
